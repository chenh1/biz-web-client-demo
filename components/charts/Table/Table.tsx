"use client";

import React, { useMemo } from 'react';
import getGapSizeClasses from '../../utils/getGapSizeClasses';
import type { IncrementKeys } from '@/components/constants/types';

interface TableData {
  value: any;
  xCell: number;
  yCell: number;
}

interface HeaderCellProps {
  children: React.ReactNode;
  columnHeaders?: { header: React.ReactNode }[];
  rowHeaders?: { header: React.ReactNode }[];
  isColumnHeader?: boolean;
  isRowHeader?: boolean;
  rowIndex: number;
  cellIndex: number;
  numberOfCellsInRow: number;
  numberOfCellsInColumn: number;
}

interface CellProps {
  children: React.ReactNode;
  columnHeaders?: { header: React.ReactNode }[];
  rowHeaders?: { header: React.ReactNode }[];
  rowIndex: number;
  cellIndex: number;
  numberOfCellsInRow: number;
  numberOfCellsInColumn: number;
  className?: string;
  value?: string | number;
}

export type TableProps = {
  objectTableData?: TableData[];
  rowHeaders?: { header: React.ReactNode }[];
  columnHeaders?: { header: React.ReactNode }[];
  CellWrapperComponent?: React.FC<CellProps>;
  CellHeaderWrapperComponent?: React.FC<HeaderCellProps>;
  gap?: IncrementKeys;
  className?: string;
}

export const DefaultHeaderCell: React.FC<HeaderCellProps> = ({
  children,
  columnHeaders,
  rowHeaders,
  isColumnHeader,
  isRowHeader,
  rowIndex,
  cellIndex,
  numberOfCellsInRow,
  numberOfCellsInColumn
}) => {
  const computedClasses = useMemo(() => {
    let cellClasses = '';

    if (isRowHeader) {
      cellClasses += 'border-l-0';

      cellClasses += (rowIndex === 0
        ? (columnHeaders?.length || 0) > 0
          ? ''
          : ' border-t-0'
        : '');

      if (rowIndex + 1 === numberOfCellsInColumn || rowIndex === numberOfCellsInColumn) {
        cellClasses += ' border-b-0';
      }
    }

    if (isColumnHeader) {
      cellClasses += ' border-t-0';

      cellClasses += (cellIndex === 0
        ? (rowHeaders?.length || 0) > 0
          ? ''
          : ' border-l-0'
        : '');

      if (cellIndex === numberOfCellsInRow || cellIndex + 1 === numberOfCellsInRow) {
        cellClasses += ' border-r-0';
      }
    }

    return cellClasses;
  }, [isRowHeader, isColumnHeader, rowIndex, cellIndex, columnHeaders, rowHeaders, numberOfCellsInRow, numberOfCellsInColumn]);

  return (
    <div className={`-m-px grid place-items-center border dark:border-max-white border-night-black border-2 p-2 ${computedClasses}`}>
      {children}
    </div>
  );
};

// DefaultCell component
export const DefaultCell: React.FC<CellProps> = ({
  children,
  columnHeaders,
  rowHeaders,
  rowIndex,
  cellIndex,
  numberOfCellsInRow,
  numberOfCellsInColumn,
  className = ""
}) => {
  const computedClasses = useMemo(() => {
    let cellClasses = '';

    if (cellIndex + 1 === numberOfCellsInRow) {
      cellClasses += ' border-r-0';
    }

    if (rowIndex + 1 === numberOfCellsInColumn) {
      cellClasses += ' border-b-0';
    }

    cellClasses += (
      rowIndex === 0
        ? (columnHeaders?.length || 0) > 0
          ? ''
          : ' border-t-0'
        : ''
    );

    cellClasses += (
      cellIndex === 0
        ? (rowHeaders?.length || 0) > 0
          ? ''
          : ' border-l-0'
        : ''
    );

    return cellClasses;
  }, [rowIndex, cellIndex, columnHeaders, rowHeaders, numberOfCellsInRow, numberOfCellsInColumn]);

  return (
    <div className={`-m-px grid place-items-center border dark:border-max-white border-night-black p-2 ${computedClasses} ${className}`}>
      {children}
    </div>
  );
};

// Table component
export const Table: React.FC<TableProps> = ({
  objectTableData = [],
  rowHeaders,
  columnHeaders,
  CellWrapperComponent = (props) => <DefaultCell {...props} />,
  CellHeaderWrapperComponent = (props) => <DefaultHeaderCell {...props} />,
  gap = "0",
  className = ""
}) => {
  const formattedTableData = useMemo(() => {
    const formattedTableData: (any[] | undefined)[] = [];

    objectTableData.forEach(({ value, xCell, yCell }) => {
      if (formattedTableData[yCell] === undefined) {
        formattedTableData[yCell] = [];
      }

      formattedTableData[yCell][xCell] = value;
    });

    return formattedTableData;
  }, [objectTableData]);

  const numberOfCellsInRow = useMemo(() => {
    return formattedTableData.reduce((acc, row) => {
      return Math.max(acc, row?.length || 0);
    }, 0);
  }, [formattedTableData]);

  const numberOfCellsInColumn = useMemo(() => {
    return formattedTableData.length;
  }, [formattedTableData]);

  const computedClasses = useMemo(() => {
    return getGapSizeClasses(gap);
  }, [gap]);

  const computedGridCols = useMemo(() => {
    const numberOfCellsInRowWithHeaders = numberOfCellsInRow + (rowHeaders?.length ? 1 : 0);
    return `repeat(${numberOfCellsInRowWithHeaders}, minmax(0, 1fr))`;
  }, [numberOfCellsInRow, rowHeaders]);

  const totalColumnHeaderCells = rowHeaders?.length === 0
    ? columnHeaders
    : [{ header: null }, ...(columnHeaders ?? [])];

  return (
    <div className={`grid ${computedClasses} ${className}`} style={{ gridTemplateColumns: computedGridCols }}>
      {formattedTableData.map((row, rowIndex) => (
        <React.Fragment key={rowIndex}>
          {rowIndex === 0 && columnHeaders && (
            (totalColumnHeaderCells ?? []).map(({ header }, i) => (
              header !== null ? (
                <CellHeaderWrapperComponent
                  key={i}
                  isColumnHeader
                  columnHeaders={columnHeaders}
                  rowHeaders={rowHeaders}
                  cellIndex={i}
                  rowIndex={rowIndex}
                  numberOfCellsInRow={numberOfCellsInRow}
                  numberOfCellsInColumn={numberOfCellsInColumn}
                >
                  {header}
                </CellHeaderWrapperComponent>
              ) : (
                <div key={i}></div>
              )
            ))
          )}
          {(row ?? []).map((cell, cellIndex) => (
            <React.Fragment key={cellIndex}>
              {cellIndex === 0 && (rowHeaders?.length || 0) > 0 && (
                <CellHeaderWrapperComponent
                  isRowHeader
                  columnHeaders={columnHeaders}
                  rowHeaders={rowHeaders}
                  rowIndex={rowIndex}
                  cellIndex={cellIndex}
                  numberOfCellsInRow={numberOfCellsInRow}
                  numberOfCellsInColumn={numberOfCellsInColumn}
                >
                  {rowHeaders?.[rowIndex]?.header}
                </CellHeaderWrapperComponent>
              )}
              <CellWrapperComponent
                value={cell}
                columnHeaders={columnHeaders}
                rowHeaders={rowHeaders}
                rowIndex={rowIndex}
                cellIndex={cellIndex}
                numberOfCellsInRow={numberOfCellsInRow}
                numberOfCellsInColumn={numberOfCellsInColumn}
              >
                {cell}
              </CellWrapperComponent>
            </React.Fragment>
          ))}
        </React.Fragment>
      ))}
    </div>
  );
};
