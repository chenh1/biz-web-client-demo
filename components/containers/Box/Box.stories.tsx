import type { BoxProps } from './Box'
import React from 'react';
import { Box } from './Box';
import { Paragraph } from '../../core/Paragraph/Paragraph';
import { Heading } from '../../core/Heading/Heading';
import { Button } from '../../core/Button/Button';

type Args = {
  args: BoxProps
}

const Template = (args: BoxProps) => (
  <Box className={"max-w-xs"} {...args}>
    <Heading type="h3">Heading text</Heading>
    <Paragraph>This is some text for a sample paragraph element</Paragraph>
    <Button label="CLICK ME" primary/>
  </Box>
)

export default {
  component: Template
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const BoxXsGap: Args = { args: { gap: "xs" } };
export const BoxSmGap: Args = { args: { gap: "sm" } };
export const BoxMdGap: Args = { args: { gap: "md" } };
export const BoxLgGap: Args = { args: { gap: "lg" } };
export const BoxXlGap: Args = { args: { gap: "xl" } };
export const BoxSeparateGap: Args = { args: { gap: "separate" } };
export const BoxDistinctGap: Args = { args: { gap: "distinct" } };

