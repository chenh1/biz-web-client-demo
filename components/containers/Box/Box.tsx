import type { IncrementKeys, PaddingValues } from '@/components/constants/types';
import React, { useMemo, RefObject, HTMLProps, ReactNode } from 'react';
import getPaddingClass from '../../utils/getPaddingClass';
import getGapSizeClasses from '../../utils/getGapSizeClasses';
import { SECTION_VIEW_CLASS } from '../../constants/tagging';

export type BoxProps = {
  innerRef?: RefObject<HTMLDivElement>; // To use with `ref` prop in React
  className?: string;
  gap?: IncrementKeys;
  padding?: IncrementKeys;
  isRounded?: boolean;
  hasShadow?: boolean;
  children?: ReactNode;
  gtmSectionView?: boolean;
  dataGtmId?: string;
  onClick?: (any) => void;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
}

const getRoundedBorder = (isRounded?: boolean): string =>
  isRounded ? 'rounded-md' : '';

const getBoxShadowClass = (hasShadow?: boolean): string =>
  hasShadow ? 'shadow-lg' : '';

export const Box: React.FC<BoxProps> = ({
  innerRef,
  className = '',
  gap = 'sm',
  padding = 'md',
  isRounded,
  hasShadow,
  children,
  gtmSectionView,
  dataGtmId,
  ...props
}) => {
  const computedGapClass = useMemo(() => {
    const gapSizeClasses = getGapSizeClasses(gap);
    const roundedBorderClass = getRoundedBorder(isRounded);
    const boxShadowClass = getBoxShadowClass(hasShadow);
    const paddingClass = getPaddingClass(padding);
    return [roundedBorderClass, gapSizeClasses, boxShadowClass, paddingClass].join(' ');
  }, [gap, isRounded, hasShadow, padding]);

  return (
    <div
      ref={innerRef}
      data-gtm-id={dataGtmId}
      className={`${gtmSectionView ? SECTION_VIEW_CLASS : ''} grid ${computedGapClass} ${className}`}
      {...props}
    >
      {children}
    </div>
  );
};
