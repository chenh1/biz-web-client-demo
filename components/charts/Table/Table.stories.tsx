import type { TableProps } from './Table'
import React from 'react';
import { Table } from './Table';
import { Box } from '../../containers/Box/Box';

type Args = {
  args: TableProps
}

export default {
  component: Table
}

export const objectTableData = [
  { xCell: 0, yCell: 0, value: "test00" },
  { xCell: 0, yCell: 1, value: "test01" },
  { xCell: 0, yCell: 2, value: "test02" },
  { xCell: 0, yCell: 3, value: "test03" },
  { xCell: 1, yCell: 0, value: "test10" },
  { xCell: 1, yCell: 1, value: "test11" },
  { xCell: 1, yCell: 2, value: "test12" },
  { xCell: 1, yCell: 3, value: "test13" },
  { xCell: 2, yCell: 0, value: "test20" },
  { xCell: 2, yCell: 1, value: "test21" },
  { xCell: 2, yCell: 2, value: "test22" },
  { xCell: 2, yCell: 3, value: "test23" },
]

export const rowHeaders = [{ header: "row1"}, { header: "row2"}, { header: "row3"}, { header: "row4"}];
export const columnHeaders = [{ header: "column1"}, { header: "column2"}, { header: "column3"}];

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const TableBoxCells: Args = {
  args: {
    objectTableData,
    rowHeaders,
    columnHeaders,
    CellWrapperComponent: props => <Box hasShadow isRounded {...props} />,
    CellHeaderWrapperComponent: props => <Box hasShadow isRounded {...props} />,
    gap: "md"
  },
};

export const TableDivCells: Args = {
  args: {
    objectTableData,
    rowHeaders,
    columnHeaders,
  },
}