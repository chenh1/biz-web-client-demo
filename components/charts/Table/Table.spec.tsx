import { Table } from "./Table";
import { render } from '@testing-library/react';
import { objectTableData, rowHeaders, columnHeaders } from './Table.stories'
import { Box } from '../../containers/Box/Box';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window


it('renders a Table given the dataset', async () => {
  const { asFragment } = render(
    <Table
      objectTableData={objectTableData}
      rowHeaders={rowHeaders}
      columnHeaders={columnHeaders}
      CellWrapperComponent={props => <Box hasShadow isRounded {...props} />}
      CellHeaderWrapperComponent={props => <Box hasShadow isRounded {...props} />}
      gap={"md"}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});