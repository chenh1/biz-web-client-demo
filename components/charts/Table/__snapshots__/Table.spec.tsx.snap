// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Table given the dataset 1`] = `
<DocumentFragment>
  <div
    class="grid gap-8 "
    style="grid-template-columns: repeat(4, minmax(0, 1fr));"
  >
    <div />
    <div
      cellindex="1"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="0"
    >
      column1
    </div>
    <div
      cellindex="2"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="0"
    >
      column2
    </div>
    <div
      cellindex="3"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="0"
    >
      column3
    </div>
    <div
      cellindex="0"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="0"
    >
      row1
    </div>
    <div
      cellindex="0"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="0"
      value="test00"
    >
      test00
    </div>
    <div
      cellindex="1"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="0"
      value="test10"
    >
      test10
    </div>
    <div
      cellindex="2"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="0"
      value="test20"
    >
      test20
    </div>
    <div
      cellindex="0"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="1"
    >
      row2
    </div>
    <div
      cellindex="0"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="1"
      value="test01"
    >
      test01
    </div>
    <div
      cellindex="1"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="1"
      value="test11"
    >
      test11
    </div>
    <div
      cellindex="2"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="1"
      value="test21"
    >
      test21
    </div>
    <div
      cellindex="0"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="2"
    >
      row3
    </div>
    <div
      cellindex="0"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="2"
      value="test02"
    >
      test02
    </div>
    <div
      cellindex="1"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="2"
      value="test12"
    >
      test12
    </div>
    <div
      cellindex="2"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="2"
      value="test22"
    >
      test22
    </div>
    <div
      cellindex="0"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="3"
    >
      row4
    </div>
    <div
      cellindex="0"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="3"
      value="test03"
    >
      test03
    </div>
    <div
      cellindex="1"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="3"
      value="test13"
    >
      test13
    </div>
    <div
      cellindex="2"
      class=" grid rounded-md gap-4 shadow-lg p-8 "
      columnheaders="[object Object],[object Object],[object Object]"
      numberofcellsincolumn="4"
      numberofcellsinrow="3"
      rowheaders="[object Object],[object Object],[object Object],[object Object]"
      rowindex="3"
      value="test23"
    >
      test23
    </div>
  </div>
</DocumentFragment>
`;
